CARDS-1887: Reference answers don't work in background jobs

acrowthe opened the pull request:
Convert ReferenceAnswersEditor and ComputedAnswersEditor over to ThreadResourceResolverProvider
- Allows these editors to run in the backend when ResourceResolverProvider does not have an effective session
- Required re-implementation of some FormUtil and SubjectUtil funtions as said util functions use the regular ResourceResolverFactory

Testing:
Running in PREMs runmode, create a visit form with Clinic, provider, a recognizable location and status. Save the form, wait a few seconds and verify that a new OED or OAIP form is automatically created with the location referenced in the new form.

Create a new patient + visit with a visit information form. Only fill out location so forms are not automatically created. Manually create a new OED or OAIP form for this visit. Observe that the location field is copied over to the hospital named below field.

Reference test forms can be found in [CARDS-1103](https://github.com/data-team-uhn/cards/pull/1116)

Verify that computed questions work correctly in the backend (tests are available in [CARDS-1800](https://github.com/data-team-uhn/cards/pull/1066)

[CARDS-1103]: https://phenotips.atlassian.net/browse/CARDS-1103?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[CARDS-1800]: https://phenotips.atlassian.net/browse/CARDS-1800?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

acrowthe committed:
CARDS-1887: Reference answers don't work in background jobs
Convert ReferenceAnswersEditor and ComputedAnswersEditor over to ThreadResourceResolverProvider
- Allows these aditors to run in the backend when ResourceResolverProvider does not have an effective session
- Required re-implementation of some Util funtions as said util functions use the regular ResourceResolverFactory

acrowthe committed:
CARDS-1887: Reference answers don't work in background jobs
Fix backend computed question regression

IntegralProgrammer commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/ComputedAnswersEditor.java:
```diff
@@ -217,12 +220,12 @@ private Map<String, Object> getNodeAnswers(final NodeState currentNode)
         final Map<String, Object> currentAnswers = new HashMap<>();
         if (currentNode.exists()) {
-            if (this.formUtils.isAnswerSection(currentNode) || this.formUtils.isForm(currentNode)) {
+            if (this.isAnswerSection(currentNode) || this.formUtils.isForm(currentNode)) {
```
How does this work? The `currentNode` object is of type `NodeState`. If `this.isAnswerSection(currentNode)` returns `false` and `this.formUtils.isForm(currentNode)` is executed, wouldn't that execution fail as `FormUtils` would attempt to use it's own session obtained from its own `ResourceResolverFactory`? Wouldn't we want to use the newly defined `this.isFormNode(NodeState node)` method instead?

IntegralProgrammer commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/AnswersEditor.java:
```diff
@@ -287,9 +281,9 @@ protected Map<String, List<NodeBuilder>> getChildNodesByReference(final NodeBuil
             String childIdentifier;
-            if (this.formUtils.isAnswerSection(childNode)) {
+            if (this.isAnswerSection(childNode)) {
                 childIdentifier = this.formUtils.getSectionIdentifier(childNode);
```
This would generate the following chain of calls; `this.formUtils.getSectionIdentifier(final NodeBuilder answerSection)` --> `this.formUtils.getSectionIdentifier(final NodeState answerSection)` --> `this.formUtils.isAnswerSection(final NodeState node)`. `this.formUtils.isAnswerSection` uses a _JCR Session_ obtained from the _ResourceResolverFactory_ referenced in https://github.com/data-team-uhn/cards/blob/80c279f2c3bbc57d2b68ebdfe9fcc6d9ebe3b764/modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/FormUtilsImpl.java#L62 instead of `this.serviceSession`.

IntegralProgrammer commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/AnswersEditor.java:
```diff
@@ -287,9 +281,9 @@ protected Map<String, List<NodeBuilder>> getChildNodesByReference(final NodeBuil
                 childIdentifier = this.formUtils.getSectionIdentifier(childNode);
-            } else if (this.formUtils.isAnswer(childNode)) {
+            } else if (this.isAnswer(childNode)) {
                 childIdentifier = this.formUtils.getQuestionIdentifier(childNode);
```
This would generate the following chain of calls; `this.formUtils.getQuestionIdentifier(final NodeBuilder answer)` --> `this.formUtils.getQuestionIdentifier(final NodeState answer)` --> `this.formUtils.isAnswer(final NodeState node)`. `this.formUtils.isAnswer` uses a _JCR Session_ obtained from the _ResourceResolverFactory_ referenced in https://github.com/data-team-uhn/cards/blob/80c279f2c3bbc57d2b68ebdfe9fcc6d9ebe3b764/modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/FormUtilsImpl.java#L62 instead of `this.serviceSession`.

IntegralProgrammer commented:
Based on reading the specifications on JIRA, I don't think that any _FormUtil_ or _SubjectUtil_ functions need to be re-implemented. From how I understand it, the JIRA specifications call for the `ResourceResolverFactory` in `FormUtilsImpl.java` to be replaced with a `ThreadResourceResolverProvider`. Once that is done, all background tasks (such as the `ComputedAnswersEditor` and the `ReferenceAnswersEditor`) would need to `push()` their `ServiceResourceResolver`s to the `ThreadResourceResolverProvider` so that it is available to `FormUtilsImpl` and then `pop()` it from `ServiceResourceResolver` when it is done being used.

An example of this in our code base can be found in the _PROMs Import Task_ https://github.com/data-team-uhn/cards/blob/80c279f2c3bbc57d2b68ebdfe9fcc6d9ebe3b764/proms-resources/backend/src/main/java/io/uhndata/cards/proms/internal/importer/ImportTask.java#L217-L226 where the `ServiceResourceResolver` is _pushed_ to the `ThreadResourceResolverProvider` after it is obtained. Obtaining a `ThreadResourceResolver` can be done using the `ThreadResourceResolverProvider.getThreadResourceResolver()` method as shown below. https://github.com/data-team-uhn/cards/blob/80c279f2c3bbc57d2b68ebdfe9fcc6d9ebe3b764/modules/form-completion-status/src/main/java/io/uhndata/cards/formcompletionstatus/AnswerCompletionStatusEditorProvider.java#L65

sashaandjic reviewed (approved):
tested

acrowthe committed:
CARDS-1887: Reference answers don't work in background jobs
- Update utility classes to use ThreadResourceResolverProviders

acrowthe committed:
CARDS-1887: Reference answers don't work in background jobs
Clean up reference and computed answer editors now that utility funtions
  properly use ThreadResourceResolverProviders

veronikaslc commented:
Tested.

sashaandjic reviewed (commented):
tested

QUESTION: all works as expected, however modifying the location in the Visit information form would not render the new location in the previously created forms. I'm not sure if this is a valid user case.

marta- commented:
> QUESTION: all works as expected, however modifying the location in the Visit information form would not render the new location in the previously created forms. I'm not sure if this is a valid user case.

That is being done separately, I believe it's 1904/1937.

marta- commented:
Rebased on the latest dev. Retesting is not necessary.

marta- commented:
Rebased on the latest `dev`. Re-testing is not necessary.

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/AnswersEditor.java:
```diff
@@ -157,7 +147,19 @@ protected Node getQuestionnaire()
-            return this.serviceSession.getNodeByIdentifier(questionnaireId);
+            Session serviceSession = null;
+            if (this.rrf != null) {
+                try (ResourceResolver serviceResolver = this.rrf.getServiceResourceResolver(
```
This isn't good. The `try-with-resource` will automatically close the created `ResourceResolver`, and thus the `serviceSession`, so the call to `getNodeByIdentifier` outside the try will fail because the session is closed.

sdumitriu replied:
The actual bug that this causes (I think, I didn't actually test yet): when the patient edits through the patient portal, in PROMS mode scores are no longer properly computed.

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/AnswersEditor.java:
```diff
@@ -157,7 +147,19 @@ protected Node getQuestionnaire()
+                try (ResourceResolver serviceResolver = this.rrf.getServiceResourceResolver(
+                    Collections.singletonMap(ResourceResolverFactory.SUBSERVICE, this.getServiceName())))
+                {
+                    serviceSession = serviceResolver == null ? null : serviceResolver.adaptTo(Session.class);
```
If you're inside the try, then `serviceResolver` is definitely not `null`; the `try` is guaranteed to have worked at this point.

sdumitriu committed:
[cleanup] Improved codestyle

sdumitriu committed:
CARDS-1887: Reference answers don't work in background jobs
- Use the thread resource resolver in more places where a resource resolver is needed
- Push a resource resolver whenever a background job is executed

sdumitriu committed:
CARDS-1887: Reference answers don't work in background jobs
Fix the service session creation

sdumitriu commented on modules/data-model/forms/impl/src/main/java/io/uhndata/cards/forms/internal/ReferenceAnswersEditorProvider.java:
```diff
@@ -32,19 +33,23 @@
  * @version $Id$
  */
-@Component
+@Component(service = EditorProvider.class, immediate = true)
```
Why this change?

sdumitriu reviewed (changes requested):
Without further changes, the behavior of `ThreadResourceResolverProvider` and `ResourceResolverProvider` is the same: in background jobs, the thread resolver is missing. Every background job must push a resolver so that it is seen from the `ThreadResourceResolverProvider`.

There are some background tasks that don't yet do that: heracles export, scheduled CSV export, webhook backup export.

sdumitriu committed:
[misc] Move ThreadResourceResolverProvider into its own module, since it is needed in an early-accessible bundle and not all of cards-utils can start that early

sdumitriu commented:
This needs a thorough test.

marta- reviewed (approved):
Looks fine to me. Tested the `prems` usecase.

sashaandjic commented:
- Start cards in proms mode
- create patient and visit Info forms (make sure that surveys completed and surveys submitted is set to `No`)
- observe that survey forms are generated
- open the Visit form and observe that Surveys Completed field is set to `Yes`

When I log as a patient I get the survey review screen presented.

sdumitriu commented:
I think that's a sign that the issue has been solved. The behavior is "correct", but now CARDS-1998 is mandatory.

sdumitriu commented:
Indeed, on dev, the forms are created without the "reference" answers for location, but on this branch the forms have the correct reference answers, and are thus marked as complete.

sdumitriu commented:
Or rather "marked as not incomplete".

sashaandjic commented:
If I manually modify Surveys Completed to `No`, why do I still get the review you answers page?

sashaandjic commented:
Scores are not calculated and shown upon submission.
![image](https://user-images.githubusercontent.com/23243244/214160782-1c293184-b0aa-4c88-979b-aa16287ad7a7.png)

sashaandjic commented:
I guess this strange behavior is also expected: As an admin I only opened the first two surveys and saved them without editing. Then I loged as a patient, and these 2 forms were shown as completed?
![image](https://user-images.githubusercontent.com/23243244/214162754-f0979ca3-3bf0-47cc-b46b-d3abf92d1acc.png)

sdumitriu commented:
Testing the patient workflow doesn't make sense until CARDS-1998 is also implemented.

acrowthe commented:
Sergiu's changes look good to me
